File: programmers_방금그곡.py
class Music:
    def __init__(self,string):
        start,end,name,scale = string.split(',')
        start_hour,start_minute = start.split(':')
        end_hour,end_minute = end.split(':')
        self.start = int(start_hour)*60 + int(start_minute)
        self.end = int(end_hour)*60 + int(end_minute)
        self.during = self.end - self.start
        self.name = name
        self.scale = []
        i = 0
        j = 0
        while i < self.during:
            i += 1
            self.scale.append(scale[j%len(scale)])
            j += 1
            if scale[j%len(scale)] == '#':
                self.scale[-1] += '#'
                j += 1

def isSong(mine,scale):
    i = 0
    mine_string = ''.join(mine)
    while i < len(scale):
        if mine[0] != scale[i]:
            i += 1
        else:
            if mine_string == ''.join(scale[i:i+len(mine)]):
                return True
            i += 1
    return False

def solution(m, musicinfos):
    scale_list = []
    for i in m:
        if i == '#':
            scale_list[-1] += i
        else:
            scale_list.append(i)
    m = scale_list[:]
    answer = '(None)'
    answer_time = -1
    start_time = -1
    for i in range(len(musicinfos)):
        musicinfos[i] = Music(musicinfos[i])
    for music in musicinfos:
        if isSong(m,music.scale):
            if answer_time < music.during:
                answer_time = music.during
                answer = music.name
                start_time = music.start
            elif answer_time == music.during:
                if start_time > music.start:
                    start_time = music.start
                    answer_time = music.during
                    answer = music.name
                    start_time = music.start
    return answer

File: programmers_방금그곡.js
class Music {
  constructor(information) {
    const [startTime, endTime, musicTitle, scale] = information.split(",");
    const [startHour, startMinute] = startTime.split(":");
    const [endHour, endMinute] = endTime.split(":");
    this.startTime = Number(startHour) * 60 + Number(startMinute);
    this.endTime = Number(endHour) * 60 + Number(endMinute);
    this.musicDuring = this.endTime - this.startTime;
    this.title = musicTitle;
    this.scale = [];
    let i = 0;
    let j = 0;
    while (i < this.musicDuring) {
      i++;
      this.scale.push(scale[j % scale.length]);
      j++;
      if (scale[j % scale.length] == "#") {
        this.scale[this.scale.length - 1] += "#";
        j++;
      }
    }
  }
}

function isSong(mine, scale) {
  let i = 0;
  let mine_string = mine.join("");
  while (i < scale.length) {
    if (mine[0] != scale[i]) i += 1;
    else {
      if (mine_string == scale.slice(i, i + mine.length).join("")) return true;
      i += 1;
    }
  }
  return false;
}

function solution(m, musicinfos) {
  const scaleList = [];
  for (var i = 0; i < m.length; i++) {
    if (i === "#") scaleList[scaleList.length - 1] += i;
    else scaleList.push(m[i]);
  }
  for (var i = 0; i < musicinfos.length; i++) {
    musicinfos[i] = new Music(musicinfos[i]);
  }
  const mArray = scaleList.slice();
  let answer = "(None)";
  let answerTime = -1;
  let startTime = -1;

  musicinfos.forEach((item) => {
    if (isSong(mArray, item.scale)) {
      if (answerTime < item.musicDuring) {
        answerTime = item.musicDuring;
        answer = item.title;
        startTime = item.startTime;
      } else if (answerTime == item.musicDuring) {
        if (startTime > item.startTime) {
          answerTime = item.musicDuring;
          answer = item.title;
          startTime = item.startTime;
        }
      }
    }
  });
  return answer;
}
